Introduce definitions and simplify structure (#11)
AMWA-TV/is-14: docs/Backup & restore.md

# Backup & restore

> Supporting backup & restore is a key feature of IS-14. In addition to supporting full back and full restore to the same device, IS-14 enables partial backup and partial restore to the same device or a different device. To discuss the various possible combinations of backup and restore, this section utilizes terms explained in the [Definitions](#definitions) section.

Reviewer: I feel we need to define more rigorously the terms "partial backup" and "partial restore".  

For instance there could be confusion over whether, when you take a partial backup and restore all of it to a Device Model, whether that is a full restore of a partial backup, or whether it's a partial restore of a partial backup.  And does the partial restore definition apply if we can only restore parts of the full backup etc.

Perhaps adding something in the Definitions section to stop that sort of confusion?

// TBD: Do we think some devices may need to reboot in order to apply a restore?

## 3. Restoring on a device with an incompatible revision

Restoring follows a similar [workflow](#2-restoring-on-a-device-with-a-compatible-revision) with the difference that the restore device does not have a compatible revision.

| ![Restoring on a different device revision](images/restoring-same-dev-different-dev-rev.png) |
|:--:|
| _**Restoring on a different device revision**_ |

Devices MUST allow the partial restoration of backups which have at least one role path `status` of `Ok` when supplying the `allowPartial` argument of `true` in the request.

// TBD: Facilities can have a multitude of instances of the same device type. Backups performed on a device instance can be used to bootstrap other instances of the same device type. What if the new device overwrites things like IP addresses? Are we saying the restore should be used as a bootstrap mechanism (do we need a isTemplate boolean flag when restoring)?

## Definitions

> A `device`, for the purposes of this section, is a physical or logical entity that can be backed up and restored using the procedures described. It may or may not correspond to an IS-04 Device or IS-04 Node.

Reviewer: Remove normative language usage.
```suggestion
A `device`, for the purposes of this section, is a physical or logical entity that can be backed up and restored using the procedures described. It is not necessarily the same entity as an IS-04 Device or IS-04 Node, but it can be.
```

> `Backup data set` is the set of data retrieved from a device using the backup procedures described.

Reviewer: Should we define this more rigorously?
```suggestion
`Backup data set` is the `NcBulkValuesHolder` object retrieved from a device using the backup procedures described.
```

> `Device revision` represents any combination of software versions (this includes firmware) and hardware revision that dictates the functionality of the device. As a vendor evolves a product through its lifecycle the hardware, software and/or firmware that make up a device is likely to change. From time to time a vendor my issue new software or firmware that a customer can install that changes how a device behaves. A device may also have hardware such a peripherals or plug-in cards added or removed that change its behaviour. Each of these changes can affect how easily a `backup data set` can be restored to a device. The `device revision` is used to represent the ensemble of the versions of all the various components that affect a device’s functionality. It is up to a vendor to decide what changes to a device represent a change of `device revision`. A software upgrade will likely result in a device being considered to have a different `device revision` but a vendor may decide that a change in, for example, PCB colour does not represent a different `device revision`.

Reviewer: Remove normative language.
```suggestion
`Device revision` represents any combination of software versions (this includes firmware) and hardware revision that dictates the functionality of the device. As a vendor evolves a product through its lifecycle the hardware, software and/or firmware that make up a device is likely to change. From time to time customers install new software or firmware issued by the vendor that changes how a device behaves. Changes to a device's hardware, such a peripherals or plug-in cards added or removed, could also change its behaviour. Each of these changes can affect how easily a `backup data set` is restored to a device. The `device revision` is used to represent the ensemble of the versions of all the various components that affect a device’s functionality. It is up to a vendor to decide what changes to a device represent a change of `device revision`. A software upgrade will likely result in a device being considered to have a different `device revision` but a vendor could decide that a change in, for example, PCB colour does not represent a different `device revision`.
```

> `Backup validation fingerprint` is an optional string in a `backup data set` that can be used to capture the various versions of the hardware, software and/or firmware that made up a device at the time the backup was performed. The `backup validation fingerprint` can be used by a device to help decide whether the `backup data set` being restored to it is compatible. The format of the string is defined by the vendor and is opaque to other systems. This may contain information like:

Reviewer: It seems that 'backup data set' is synonymous with `NcBulkValuesHolder`.  Should we converge on one or the other as the preferred terminology? Perhaps rename `NcBulkValuesHolder` to `NcBackupDataSet` to make it more obvious?

Author: We are continuing to use `backup data set`.

> `Backup validation fingerprint` is an optional string in a `backup data set` that can be used to capture the various versions of the hardware, software and/or firmware that made up a device at the time the backup was performed. The `backup validation fingerprint` can be used by a device to help decide whether the `backup data set` being restored to it is compatible. The format of the string is defined by the vendor and is opaque to other systems. This may contain information like:

Reviewer: Remove normative language.
```suggestion
`Backup validation fingerprint` is an optional string in a `backup data set` that can be used to capture the various versions of the hardware, software and/or firmware that made up a device at the time the backup was performed. The `backup validation fingerprint` can be used by a device to help decide whether the `backup data set` being restored to it is compatible. The format of the string is defined by the vendor and is opaque to other systems. This could contain information such as:
```

- Manufacturer key
- Product key
- Software versions
- Hardware revisions
- Backup response hash
- Timestamp
- Whether its a full device model backup or a subset

> 

Reviewer: Some strawman definitions for discussion:

```suggestion

A `full backup` is a `backup data set` returned by a `GET` operation on the `/bulkProperties` endpoint of the root block with the `recurse` parameter set to true.

A `partial backup` is a `backup data set` returned by a `GET` operation on the `/bulkProperties` endpoint of any object of the Device Model, other than the root block. A partial backup can also be obtained by a `GET` operation on the `/bulkProperties` endpoint of the root block with the `recurse` query parameter set to false; however this will only result in a backup of the root block only.

A `full restore` is when an `backup data set` from a `full backup` is applied as a `PUT` to the `/bulkProperties` endpoint of the root block of the Device Model.

A `partial restore` is when an `backup data set` from a `partial backup` is applied as a `PUT` to the `/bulkProperties` endpoint of the object of the Device Model from which the `partial backup` originated.
```

Author: We also have the `recurse` flag as an argument in SetPropertiesByPath. This should be reflected in the definitions of `full restore` and `partial restore`

> A `compatible revision` is a change of `device revision` such that all of the `backup data set` taken before the revision can be successfully validated by the modified device.

Reviewer: Good to be clear on `incompatible revision` also.
```suggestion
A `compatible revision` is a change of `device revision` such that all of the `backup data set` taken before the revision can be successfully validated by the modified device. If the `backup data set` can not be successfully validated it is said to be a `incompatible revision`.
```
